File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.java
class Solution {
    public long gcd(long a, long b){
        if(b == 0) return a;

        return gcd(b ,a % b);
    }
    public long gcdSum(int[] nums) {
        long n = nums.length;
        
        long[] prefix = new long[(int)n];
        long[] max = new long[(int)n];
        max[0] = nums[0];

        for(int i = 1; i < n; i++){
            max[i] = Math.max(nums[i],max[i-1]);
        }
        
        for(int i = 0; i < (int)n; i++){
            prefix[i] = gcd(max[i],nums[i]);
        }

        Arrays.sort(prefix);

        long sum = 0;

        int i = 0;
        int j = prefix.length-1;

        while(i < j){
            sum += gcd(prefix[i],prefix[j]);
            i++;
            j--;
        }

        return sum;
    }
}

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.py

class Solution:
    def gcdSum(self, nums: list[int]) -> int:
        amax = []
        amax.append(nums[0])

        for i in range(1,len(nums)):
            amax.append(max(nums[i],amax[i-1]))

        prefix = [math.gcd(a, n) for a, n in zip(amax, nums)]

        prefix.sort()

        i = 0
        j = len(prefix)-1

        sum = 0

        while(i < j):
            sum += (math.gcd(prefix[i],prefix[j]))
            i += 1
            j -= 1

        return sum
        